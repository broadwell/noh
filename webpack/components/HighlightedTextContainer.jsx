import React from "react";
import PropTypes from "prop-types";
import { connect } from "react-redux";
import HighlightedText from "./HighlightedText";
import store from "../store";
import { setCurrentPhraseID } from "../actionCreators";

function setCurrentPhrase(idString) {
  store.dispatch(setCurrentPhraseID(idString));
}

class HighlightedTextContainer extends React.Component {
  componentWillReceiveProps() {
    const currentPhraseID = this.determineCurrentPhrase();
    setCurrentPhrase(currentPhraseID);
  }

  determineCurrentPhrase() {
    let currentPhraseID = "";
    for (let i = 0; i < this.props.phrases.length; i += 1) {
      if (
        this.props.currentTime >= this.props.phrases[i].startTime &&
        this.props.currentTime <= this.props.phrases[i].endTime
      ) {
        currentPhraseID = this.props.phrases[i].phraseID;
      }
    }
    return currentPhraseID;
  }

  render() {
    return (
      <div>
        <p>Current phraseID: {this.props.currentPhraseID}</p>
        <p>Current time: {this.props.currentTime}</p>
        <HighlightedText
          singingStyle={this.props.singingStyle}
          phrases={this.props.phrases}
          currentPhraseID={this.props.currentPhraseID}
        />
      </div>
    );
  }
}

HighlightedTextContainer.propTypes = {
  currentTime: PropTypes.number.isRequired,
  singingStyle: PropTypes.string.isRequired,
  currentPhraseID: PropTypes.string.isRequired,
  phrases: PropTypes.arrayOf(
    PropTypes.shape({
      phraseID: PropTypes.string,
      startTime: PropTypes.number,
      endTime: PropTypes.number,
      transcription: PropTypes.string,
      translation: PropTypes.string
    })
  ).isRequired
};

const mapStateToProps = state => ({
  currentTime: state.currentTime,
  currentPhraseID: state.currentPhraseID
});

export const Unwrapped = HighlightedTextContainer;
export default connect(mapStateToProps)(HighlightedTextContainer);
