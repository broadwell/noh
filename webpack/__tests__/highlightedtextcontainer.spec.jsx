import React from "react";
import { shallow, mount } from "enzyme";
import configureMockStore from "redux-mock-store";
import HighlightedTextContainer, {
  Unwrapped as UnwrappedHighlightedTextContainer
} from "../components/HighlightedTextContainer";

const phrases = [
  {
    phraseID: "I/1",
    startTime: 0,
    endTime: 10,
    transcription: "transcription line 1",
    translation: "translation line 1"
  },
  {
    phraseID: "I/2",
    startTime: 11,
    endTime: 20,
    transcription: "transcription line 2",
    translation: "translation line 2"
  }
];

describe("<UnwrappedHighlightedTextContainer>", () => {
  it("renders as expected unconnected to redux", () => {
    const setCurrentPhrase = jest.fn();
    const component = shallow(
      <UnwrappedHighlightedTextContainer
        currentTime={13}
        singingStyle="spoken"
        phrases={phrases}
        currentPhraseID="I/1"
        setCurrentPhrase={setCurrentPhrase}
      />
    );
    expect(component).toMatchSnapshot();
  });
  it("correctly determines current phrase id", () => {
    const setCurrentPhrase = jest.fn();
    const component = shallow(
      <UnwrappedHighlightedTextContainer
        currentTime={13}
        singingStyle="spoken"
        phrases={phrases}
        currentPhraseID="I/1"
        setCurrentPhrase={setCurrentPhrase}
      />
    );
    const calculatedPhraseID = component.instance().determineCurrentPhrase();
    expect(calculatedPhraseID).toBe("I/2");
  });
});

describe("HighlightedTextContainer", () => {
  it("dispatches the correct action on componentWillReceiveProps", () => {
    const initialState = { currentTime: 13, currentPhraseID: "" };
    const mockStore = configureMockStore();
    const store = mockStore(initialState);
    const wrapper = mount(
      <HighlightedTextContainer singingStyle="spoken" phrases={phrases} />,
      { context: { store } }
    );
    wrapper.setProps({ currentTime: 16 });
    const action = { type: "SET_CURRENT_PHRASE_ID", payload: "I/2" };
    expect(store.getActions()[0]).toEqual(action);
  });
});
