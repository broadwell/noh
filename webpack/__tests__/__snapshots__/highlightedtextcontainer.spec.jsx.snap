// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<UnwrappedHighlightedTextContainer> renders as expected unconnected to redux 1`] = `
<div>
  <HighlightedText
    currentPhraseID="I/1"
    phrases={
      Array [
        Object {
          "endTime": 10,
          "phraseID": "I/1",
          "startTime": 0,
          "transcription": "transcription line 1",
          "translation": "translation line 1",
        },
        Object {
          "endTime": 20,
          "phraseID": "I/2",
          "startTime": 11,
          "transcription": "transcription line 2",
          "translation": "translation line 2",
        },
      ]
    }
    singingStyle=""
  />
</div>
`;
